### PR TITLE
feature: mempool v1 implementation

### PR DESCRIPTION
**Pull request by alexanderbez**
Implementation of [ADR-067](https://github.com/tendermint/tendermint/blob/master/docs/architecture/adr-067-mempool-refactor.md) -- mempool v1.

Note, to aid in review, I've decided to split up this work into two PRs:

1. This PR, which handles the foundation of moving parts around to support versioned mempool reactors. Note, there are no actual protocol or software changes in this PR -- just moving things around.
2. The mempool v1 reactor implementation. This PR (https://github.com/tendermint/tendermint/pull/6468) contains the actual changes outlined in [ADR-067](https://github.com/tendermint/tendermint/blob/master/docs/architecture/adr-067-mempool-refactor.md).

### CODE REVIEW

**Review by marbar3778 — APPROVED** [review]
Utack

Need a change log entry for the changes

File: config/config.go [hunk 1]
```diff
@@ -741,9 +745,10 @@ type MempoolConfig struct {
 	MaxBatchBytes int `mapstructure:"max-batch-bytes"`
 }
 
-// DefaultMempoolConfig returns a default configuration for the Tendermint mempool
+// DefaultMempoolConfig returns a default configuration for the Tendermint mempool.
 func DefaultMempoolConfig() *MempoolConfig {
 	return &MempoolConfig{
+		Version:   MempoolV1,
```
Comment (cmwaters) [inline config/config.go:751]: Don't we want the default to be `MempoolV0` at least until V1 is complete else the node is just going to panic for anything that uses the default config

**Review by cmwaters — APPROVED** [review]
LGTM 👍

**Comment by alexanderbez** [comment]
> Need a change log entry for the changes

That will come in the 2nd PR.

File: config/config.go (hunk 1 above)
Comment (alexanderbez) [reply config/config.go:751]: No, we want v1 to be default. Once I wrap up the 2nd PR, nothing will panic or fail.

Comment (cmwaters) [reply config/config.go:751]: Ok. So do you plan to merge the PR's together before merging on master?

Comment (alexanderbez) [reply config/config.go:751]: Yes, exactly :)

**Comment by tychoish** [comment]
This seems fine, though I wonder if it's necessary:

- the new mempool, passed with default (zero values) for the new chectx should have the same behavior and semantics as the legacy mempool. 
- adding a configuration option may be good as a back door, if there are any unforeseen problems with the new code at scale, but I think long term (certainly, by 1.0) it seems like it would be better to just have the one implementation (testing overhead, maintenance/support burden)

**Comment by alexanderbez** [comment]
@tychoish,

> the new mempool, passed with default (zero values) for the new chectx should have the same behavior and semantics as the legacy mempool.

I'm sorry, but I don't follow this...CheckTx will perform the same logic. Wdym by passing values? 

> adding a configuration option may be good as a back door, if there are any unforeseen problems with the new code at scale, but I think long term (certainly, by 1.0) it seems like it would be better to just have the one implementation (testing overhead, maintenance/support burden)

It depends. There could be multiple mempools we maintain (as @ValarDragon pointed out) and various operator types may want to use different ones, e.g. validators use X and and sentries/full-nodes use Y, etc...That being said, I see no reason why we'd maintain v0 after 0.35 or after.

**Comment by tychoish** [comment]
> I'm sorry, but I don't follow this...CheckTx will perform the same logic. Wdym by passing values?

My assertion/understanding is that using the new mempool with an old application (e.g. one that passes 0 priority for everything will have the same semantics, and be isomorphic to the legacy mempool. If this is true, is there any reason to let the mempool be configurable? 

Similarly, particularly since we're not likely to maintain the legacy implementation in the long term, is providing configuration in the short term going to be valuable, and not just create more operational pain when we remove it/etc. Though it's definitely true that I don't 100% understand the operational case for using the legacy mempool in some cases?

**Comment by alexanderbez** [comment]
> My assertion/understanding is that using the new mempool with an old application (e.g. one that passes 0 priority for everything will have the same semantics, and be isomorphic to the legacy mempool. If this is true, is there any reason to let the mempool be configurable?

Gossiping remains the same, however, block inclusion/construction will not be the same.

**Pull request merged**